src/fractals/newtons_method.rs

```diff
@@ -0,0 +1,372 @@
+    pub soln: Complex64,
+
+    /// Number of iterations taken to converge. In range `[0, max_iteration_count]` inclusive.
+    pub iteration_count: u32,
+
+    /// A smooth iteration count, used for rendering. It is computed based on the quadratic
+    /// convergence behavior of the Newton-Rhapson method near a fixed point.
+    pub smooth_iteration_count: f32,
+}
+
+/// Returns Some(NewtonRhapsonResult) if the iteration converges within
+/// `max_iteration_count` iterations to within `convergence_tolerance`. Otherwise returns None.
+pub fn newton_rhapson_iteration_sequence<F: ComplexFunctionWithSlope>(
+    system: &F,
+    z0: Complex64,
+    convergence_tolerance: f64,
+    max_iteration_count: u32,
+) -> Option<NewtonRhapsonResult> {
+    let mut z_prev = z0;
+    let mut prev_err: Option<f64> = None;
+
+    for iteration in 0..=max_iteration_count {
+        let z_next = system.newton_rhapson_step(z_prev);
+        let error = (z_next - z_prev).norm_sqr();
+
+        if error < convergence_tolerance {
+            let iteration_count = iteration;
+
+            let smooth_iteration_count = if let Some(e_prev) = prev_err {
+                // model error as geometric between e_prev and err
+                let error_ratio = error / e_prev;
+
+                // Guard against the case where error actually hits zero, which would cause ln(0).
+                if error > 0.0 {
+                    // Model the error as geometric between the last two steps:
+                    //   e_n ≈ e_prev * error_ratio^(n - (k - 1))
+                    // Solve e_ν = tol for the fractional iteration index ν:
+                    //   ν = (k - 1) + ln(tol / e_prev) / ln(error_ratio)
+                    let frac = (convergence_tolerance / e_prev).ln() / error_ratio.ln();
+                    ((iteration as f64 - 1.0) + frac) as f32
+                } else {
+                    iteration as f32
+                }
+            } else {
+                // We converged on the first iteration, so we have no data for interpolation.
+                iteration as f32
+            };
+
+            return Some(NewtonRhapsonResult {
+                soln: z_next,
+                iteration_count,
+                smooth_iteration_count,
+            });
+        }
+
+        prev_err = Some(error);
+        z_prev = z_next;
+    }
+
+    // Only reach here if we fail to converge.
+    None
+}
+
+// Used to interpolate between two color values based on the iterations
+// required for the Newton-Raphson method to converge to a root.
+// Query values of 0 map to `iteration_limits[0]` and values of 1 map to
+// `max_iteration_count`. The `value` of zero corresponds to the common
+// background color, while a `value` of one corresponds to the foreground
+// color associated with the root that the iteration converges to.
+#[derive(Serialize, Deserialize, Debug, Clone, Copy)]
+pub struct GrayscaleMapKeyFrame {
```

Comment:
I think we can do better than this eventually, but this is good enough for the simple first example. I want to eventually support arbitrary color maps for each basin of attraction.